File: SC/Assignment3.cpp
#include <iostream>   // for input/output
#include <vector>     // to use vectors (dynamic arrays)
#include <cmath>      // for fabs(), pow(), etc.
#include <cstdlib>    // for rand()
#include <ctime>      // for seeding random generator
#include <limits>     // for infinity (to initialize best scores)
using namespace std;

// Objective function: Sphere function (minimization problem)
// f(x) = sum(x_i^2), global minimum is at (0,0,...,0) with value 0
double objective(const vector<double>& position) {
    double sum = 0.0;
    for (double x : position) sum += x * x; // sum of squares
    return sum;
}

// Generate random double between [min, max]
double randDouble(double min, double max) {
    return min + (max - min) * ((double)rand() / RAND_MAX);
}

int main() {
    srand(time(0));  // seed random number generator with current time

    int dim = 5;                // dimension of problem (number of variables)
    int wolves = 20;            // number of wolves in pack
    int maxIter = 100;          // maximum number of iterations
    double lb = -10.0, ub = 10.0; // lower and upper bounds of search space

    // Initialize positions of wolves randomly in search space
    vector<vector<double>> positions(wolves, vector<double>(dim));
    for (int i = 0; i < wolves; i++)
        for (int d = 0; d < dim; d++)
            positions[i][d] = randDouble(lb, ub);

    // Alpha, Beta, Delta wolves (top 3 best solutions found so far)
    vector<double> Alpha(dim), Beta(dim), Delta(dim);
    double AlphaScore = numeric_limits<double>::infinity(); // best fitness
    double BetaScore  = numeric_limits<double>::infinity(); // 2nd best
    double DeltaScore = numeric_limits<double>::infinity(); // 3rd best

    // ================= MAIN LOOP ==================
    for (int iter = 0; iter < maxIter; iter++) {
        // Evaluate all wolves
        for (int i = 0; i < wolves; i++) {
            double fitness = objective(positions[i]);

            // Update Alpha, Beta, Delta based on fitness
            if (fitness < AlphaScore) {
                DeltaScore = BetaScore; Delta = Beta;   // shift down
                BetaScore = AlphaScore; Beta = Alpha;
                AlphaScore = fitness; Alpha = positions[i]; // new best
            }
            else if (fitness < BetaScore) {
                DeltaScore = BetaScore; Delta = Beta;
                BetaScore = fitness; Beta = positions[i]; // new 2nd best
            }
            else if (fitness < DeltaScore) {
                DeltaScore = fitness; Delta = positions[i]; // new 3rd best
            }
        }

        // 'a' decreases linearly from 2 to 0, controls exploration/exploitation
        double a = 2.0 - iter * (2.0 / maxIter);

        // Update position of each wolf
        for (int i = 0; i < wolves; i++) {
            for (int d = 0; d < dim; d++) {
                // Influence of Alpha wolf
                double r1 = randDouble(0,1), r2 = randDouble(0,1);
                double A1 = 2*a*r1 - a;   // coefficient A
                double C1 = 2*r2;         // coefficient C
                double D_alpha = fabs(C1*Alpha[d] - positions[i][d]); // distance to alpha
                double X1 = Alpha[d] - A1*D_alpha; // candidate position wrt alpha

                // Influence of Beta wolf
                r1 = randDouble(0,1); r2 = randDouble(0,1);
                double A2 = 2*a*r1 - a;
                double C2 = 2*r2;
                double D_beta = fabs(C2*Beta[d] - positions[i][d]);
                double X2 = Beta[d] - A2*D_beta;

                // Influence of Delta wolf
                r1 = randDouble(0,1); r2 = randDouble(0,1);
                double A3 = 2*a*r1 - a;
                double C3 = 2*r2;
                double D_delta = fabs(C3*Delta[d] - positions[i][d]);
                double X3 = Delta[d] - A3*D_delta;

                // New position = average influence of Alpha, Beta, Delta
                positions[i][d] = (X1 + X2 + X3) / 3.0;

                // Keep the wolf inside search boundaries
                if (positions[i][d] < lb) positions[i][d] = lb;
                if (positions[i][d] > ub) positions[i][d] = ub;
            }
        }

        // Print best fitness found so far (progress)
        cout << "Iteration " << iter+1 << " best fitness = " << AlphaScore << endl;
    }

    // ================= FINAL RESULT ==================
    cout << "\nBest solution found (Alpha wolf): ";
    for (double x : Alpha) cout << x << " "; // print coordinates of best wolf
    cout << "\nBest objective value = " << AlphaScore << endl;

    return 0; // program ends successfully
}

File: SC/Assignment3.java
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class Assignment3 {

    // Objective function: Sphere -> sum(x^2)
    static double objective(ArrayList<Double> position) {
        double sum = 0.0;
        for (double x : position) sum += x * x;
        return sum;
    }

    // Random double in range [min, max]
    static double randDouble(Random rand, double min, double max) {
        return min + (max - min) * rand.nextDouble();
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        Random rand = new Random();

        System.out.print("Enter number of dimensions: ");
        int dim = sc.nextInt();

        System.out.print("Enter number of wolves: ");
        int wolves = sc.nextInt();

        System.out.print("Enter max iterations: ");
        int maxIter = sc.nextInt();

        System.out.print("Enter lower bound: ");
        double lb = sc.nextDouble();

        System.out.print("Enter upper bound: ");
        double ub = sc.nextDouble();

        // Initialize wolf positions
        ArrayList<ArrayList<Double>> positions = new ArrayList<>();
        for (int i = 0; i < wolves; i++) {
            ArrayList<Double> pos = new ArrayList<>();
            for (int d = 0; d < dim; d++)
                pos.add(randDouble(rand, lb, ub));
            positions.add(pos);
        }

        ArrayList<Double> Alpha = new ArrayList<>();
        ArrayList<Double> Beta  = new ArrayList<>();
        ArrayList<Double> Delta = new ArrayList<>();

        double AlphaScore = Double.POSITIVE_INFINITY;
        double BetaScore  = Double.POSITIVE_INFINITY;
        double DeltaScore = Double.POSITIVE_INFINITY;

        // Main Loop
        for (int iter = 0; iter < maxIter; iter++) {

            for (int i = 0; i < wolves; i++) {

                double fitness = objective(positions.get(i));

                // Rank wolves
                if (fitness < AlphaScore) {
                    DeltaScore = BetaScore;  Delta = new ArrayList<>(Beta);
                    BetaScore  = AlphaScore; Beta  = new ArrayList<>(Alpha);
                    AlphaScore = fitness;     Alpha = new ArrayList<>(positions.get(i));
                }
                else if (fitness < BetaScore) {
                    DeltaScore = BetaScore;  Delta = new ArrayList<>(Beta);
                    BetaScore  = fitness;    Beta = new ArrayList<>(positions.get(i));
                }
                else if (fitness < DeltaScore) {
                    DeltaScore = fitness;    Delta = new ArrayList<>(positions.get(i));
                }
            }

            double a = 2.0 - iter * (2.0 / maxIter);

            // Update positions
            for (int i = 0; i < wolves; i++) {
                for (int d = 0; d < dim; d++) {

                    double r1 = rand.nextDouble(), r2 = rand.nextDouble();
                    double A1 = 2 * a * r1 - a;
                    double C1 = 2 * r2;
                    double D_alpha = Math.abs(C1 * Alpha.get(d) - positions.get(i).get(d));
                    double X1 = Alpha.get(d) - A1 * D_alpha;

                    r1 = rand.nextDouble(); r2 = rand.nextDouble();
                    double A2 = 2 * a * r1 - a;
                    double C2 = 2 * r2;
                    double D_beta = Math.abs(C2 * Beta.get(d) - positions.get(i).get(d));
                    double X2 = Beta.get(d) - A2 * D_beta;

                    r1 = rand.nextDouble(); r2 = rand.nextDouble();
                    double A3 = 2 * a * r1 - a;
                    double C3 = 2 * r2;
                    double D_delta = Math.abs(C3 * Delta.get(d) - positions.get(i).get(d));
                    double X3 = Delta.get(d) - A3 * D_delta;

                    double newPos = (X1 + X2 + X3) / 3.0;

                    if (newPos < lb) newPos = lb;
                    if (newPos > ub) newPos = ub;

                    positions.get(i).set(d, newPos);
                }
            }

            System.out.println("Iteration " + (iter + 1) + " best fitness = " + AlphaScore);
        }

        System.out.print("\nBest solution found (Alpha wolf): ");
        for (double x : Alpha) System.out.print(x + " ");

       
 System.out.println("\nBest objective value = " + AlphaScore);
        sc.close();
    }
}
